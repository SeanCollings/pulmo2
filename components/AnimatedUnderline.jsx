import React, { useEffect } from 'react';
import { StyleSheet, View, Text, Animated } from 'react-native';

import { useTheme } from '../hooks/useTheme';

const AnimatedUnderline = ({ text, duration = 2000 }) => {
  const theme = useTheme();
  const animatedGrow = new Animated.Value(0);
  const animationOpacity = new Animated.Value(theme.DARK ? 0.5 : 0.7);

  const interpolateGrow = animatedGrow.interpolate({
    inputRange: [0, 1],
    outputRange: ['0%', '90%'],
  });

  const lineGrowAnimation = Animated.timing(animatedGrow, {
    toValue: 1,
    duration,
  });
  const opacityAnimation = Animated.timing(animationOpacity, {
    toValue: 0,
    duration: 1000,
  });

  useEffect(() => {
    Animated.loop(
      Animated.sequence([lineGrowAnimation, opacityAnimation])
    ).start();
  }, []);

  const opacity = theme.DARK ? 0.87 : 1;
  const animatedGrowStyle = {
    width: interpolateGrow,
    opacity: animationOpacity,
  };

  return (
    <View style={styles.container}>
      <Text style={{ ...styles.textStyle, color: theme.TEXT, opacity }}>
        {text}
      </Text>
      <View style={styles.lineContainer}>
        <View
          style={{
            ...styles.horizontalLine,
            borderBottomColor: theme.BACKGROUND,
            width: '100%',
          }}
        ></View>
        <Animated.View
          style={{
            ...styles.horizontalLine,
            borderBottomColor: theme.TEXT,
            position: 'absolute',
            ...animatedGrowStyle,
          }}
        ></Animated.View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
  },
  textStyle: {
    fontSize: 40,
    textAlignVertical: 'center',
    fontFamily: 'tit-light',
    textTransform: 'lowercase',
  },
  lineContainer: {
    width: '100%',
    alignItems: 'center',
  },
  horizontalLine: {
    borderBottomWidth: 1,
  },
});

export default AnimatedUnderline;
