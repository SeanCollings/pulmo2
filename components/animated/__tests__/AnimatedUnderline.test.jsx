import React from 'react';
import { act, create } from 'react-test-renderer';
import { cleanup } from 'react-native-testing-library';
import AnimatedUnderline from '../AnimatedUnderline';

jest.useFakeTimers();
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

describe('AnimatedUnderline - unit test', () => {
  let component;
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  test('should render default', async () => {
    await act(async () => {
      component = create(<AnimatedUnderline duration={0} fadeOut={0} jest />);
    });
    expect(component).toMatchSnapshot();
  });
});
