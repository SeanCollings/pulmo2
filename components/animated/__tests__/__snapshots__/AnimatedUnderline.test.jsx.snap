// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedUnderline - unit test should render default 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
    }
  }
>
  <View />
  <View
    style={
      Object {
        "alignItems": "center",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "borderBottomColor": "white",
          "borderBottomWidth": 1,
          "width": "100%",
        }
      }
    />
    <View
      style={
        Object {
          "borderBottomColor": "#3e4b57",
          "borderBottomWidth": 1,
          "opacity": 0.7,
          "position": "absolute",
          "width": "0%",
        }
      }
    />
  </View>
</View>
`;
